### PR TITLE
[OMSI] Preparing Code generation for C

### PR DESCRIPTION
**Pull request by AnHeuermann**
### Purpose
Preparing OpenModelica/OMCompiler master for generating OMSUs with `simCodeTarget=omsic`.

### Files Changed
- Compiler/Template/CodegenCFunctions.tpl
  - Use new C data structure for generated code for OMSUs.
  - Added function `crefOMSI` and C basic data types depending on `simCodeTarget`.
- Compiler/Template/CodegenUtil.tpl
  - Added `varKinds` `LOOP_ITERATION()` and `LOOP_SOLVED()`. Those are used for solving algebraic loops.
- Compiler/Template/CodegenUtilSimulation.tpl
  - Added function `template fullPathPrefix(SimCode simCode)` to get `fullPathPrefix` from `SimCode`. Used for generating OMSUs in specified folder.
  - Return `equationIndex` for `SES_ALGEBRAIC_SYSTEM`.
  - Added dump functions for `derivativeMatrix` and `OMSIFunction`.
- Compiler/Template/SimCodeTV.mo
  - Added optional `OMSIData` to `SimCode`
  - Added typeview for new records `OMSIData`, `SES_ALGEBRAIC_SYSTEM`, `DerivativeMatrix`, `OMSI_CONTEXT`, `LOOP_ITERATION`, `LOOP_SOLVED`
  - Added typeview for new functions `getLocalValueReference`, `localCref2SimVar`, crefPrefixPre
- Compiler/Stubs/SimCodeUtil.mo
  - Added `function localCref2SimVar<A,B>`, `function getValueReference`, `function getLocalValueReference<A>`

### New Files
- Compiler/Template/CodegenOMSIC.tpl
  - Functions for code generation only used for `simCodeTarget=omsic`
- Compiler/Template/CodegenOMSIC_Equations.tpl
  - Mainly functions for code generation for equation code. Used for C and C++
- Compiler/Template/CodegenOMSI_common.tpl
  - Shared template functions, used for `simCodeTarget=omsic` and `simCodeTarget=omsicpp`

### CODE REVIEW

File: Compiler/Template/SimCodeTV.mo [hunk 1]
```diff
@@ -395,6 +395,7 @@ package SimCode
       PartitionData partitionData;
       Option<DaeModeData> daeModeData;
       list<SimEqSystem> inlineEquations;
+      Option<OMSIData> omsiData;
```
Comment (lochel) [inline Compiler/Template/SimCodeTV.mo]: Is there redundant data that will be removed in favour of `omsiData`?

File: Compiler/Template/CodegenOMSI_common.tpl [hunk 2]
```diff
@@ -0,0 +1,835 @@
...
+ * FOR A PARTICULAR PURPOSE, EXCEPT AS EXPRESSLY SET FORTH
+ * IN THE BY RECIPIENT SELECTED SUBSIDIARY LICENSE CONDITIONS OF OSMC-PL.
+ *
+ * See the full OSMC Public License conditions for more details.
+ *
+ */
+
+ package CodegenOMSI_common
+" file:        CodegenOMSIC.tpl
+  package:     CodegenOMSIC
```
Comment (lochel) [inline Compiler/Template/CodegenOMSI_common.tpl]: Looks like copy&paste errors. I suggest removing useless comments, i.e. file and package.

**Comment by lochel** [comment]
What is an `OMSU`? I guess it stands for OpenModelica simulation unit. Is it a shared object, an FMU, or even a simulator itself?

File: Compiler/Template/SimCodeTV.mo (hunk 1 above)
Comment (AnHeuermann) [reply Compiler/Template/SimCodeTV.mo]: Definitely. My guess would be, that most (better all) of the `list<SimEqSystem>` above are redundant and should be deleted in a working version of the OMSI.

**Comment by AnHeuermann** [comment]
> What is an `OMSU`? I guess it stands for OpenModelica simulation unit.

Yep.
>  Is it a shared object, an FMU, or even a simulator itself?

It's a Model-Exchang FMU

**Comment by OpenModelica-Hudson** [comment]
The test suite is unstable according to [OpenModelica_TEST_PULL_REQUEST 2019-02-27_17-46-02](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/6474/).

**Comment by OpenModelica-Hudson** [comment]
The tests run correctly according to [OpenModelica_TEST_PULL_REQUEST 2019-02-28_14-32-30](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/6475/console).

**Comment by OpenModelica-Hudson** [comment]
The tests run correctly according to [OpenModelica_TEST_PULL_REQUEST 2019-03-07_15-06-45](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/6496/console).

**Pull request closed without merging**